File: 2289.steps-to-make-array-non-decreasing.cpp
#
# @lc app=leetcode id=2289 lang=cpp
#
# [2289] Steps to Make Array Non-decreasing
#

# @lc code=start
class Solution {
public:
    int totalSteps(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, 0);
        stack<int> st;
        int ans = 0;
        for (int i = 0; i < n; ++i) {
            int cur_steps = 0;
            while (!st.empty() && nums[st.top()] <= nums[i]) {
                cur_steps = max(cur_steps, dp[st.top()] + 1);
                st.pop();
            }
            if (!st.empty()) {
                dp[i] = max(cur_steps, 1);
            }
            ans = max(ans, dp[i]);
            st.push(i);
        }
        return ans;
    }
};
# @lc code=end

File: 2289.steps-to-make-array-non-decreasing.py
#
# @lc app=leetcode id=2289 lang=python3
#
# [2289] Steps to Make Array Non-decreasing
#

# @lc code=start
from typing import List

class Solution:
    def totalSteps(self, nums: List[int]) -> int:
        stack = []
        ans = 0
        for num in reversed(nums):
            steps = 0
            while stack and stack[-1] < num:
                steps += 1
                stack.pop()
            ans = max(ans, steps)
            stack.append(num)
        return ans

# @lc code=end